pkg/api/handlers/utils/containers.go

```diff
@@ -202,12 +203,30 @@ func waitRemoved(ctrWait containerWaitFn) (int32, error) {
 	return code, err
 }
 
-func waitNextExit(ctrWait containerWaitFn) (int32, error) {
-	_, err := ctrWait(define.ContainerStateRunning)
-	if err != nil {
-		return -1, err
+func waitNextExit(ctx context.Context, containerName string) (int32, error) {
+	runtime := ctx.Value("runtime").(*libpod.Runtime)
+	var containerEngine entities.ContainerEngine = &abi.ContainerEngine{Libpod: runtime}
+	eventChannel := make(chan *events.Event)
+	errChannel := make(chan error)
+	opts := entities.EventsOptions{
+		EventChan: eventChannel,
+		Filter:    []string{"event=died", fmt.Sprintf("container=%s", containerName)},
+		Stream:    true,
+	}
+
+	// ctx is used to cancel event watching goroutine
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		errChannel <- containerEngine.Events(ctx, opts)
+	}()
+
+	evt, ok := <-eventChannel
+	if ok {
+		return int32(evt.ContainerExitCode), nil
+	} else {
+		return -1, <-errChannel
```

Comment:
What if we get `nil` from the `errChannel`?

Since we're streaming, the container _must_ die at some point.  So if we reach line 228, somethings off and I think we need a custom error for that (see previous comment).

Reply:
>What if we get nil from the errChannel?

I thin that is all right, isn't it? It happens if the request was canceled.

>Since we're streaming, the container must die at some point.

But the container may die after request cancelation.

The line 228 can be reached if the context was canceled (e.g. client closed connection) in which case err is nil.

Reply:
we reach 228 if `ok == false` which happens if the `eventChannel` has been closed, and the channel is closed by `containerEngine.Events()` just before the function exits.

Reply:
does it make sense @vrothberg ?